java/serve/src/main/java/io/ray/serve/api/ServeControllerClient.java

```diff
@@ -0,0 +1,410 @@
+      throw new RayServeException(LogUtil.format("Endpoint {} does not exist.", endpointName));
+    }
+
+    RayServeHandle handle = new RayServeHandle(controller, endpointName, null, null);
+    handleCache.put(cacheKey, handle);
+    return handle;
+  }
+
+  public void deploy(
+      String name,
+      String deploymentDef,
+      Object[] initArgs,
+      Map<String, Object> rayActorOptions,
+      DeploymentConfig deploymentConfig,
+      String version,
+      String prevVersion,
+      String routePrefix,
+      String url,
+      Boolean blocking) {
+
+    if (deploymentConfig == null) {
+      deploymentConfig = new DeploymentConfig();
+    }
+    if (rayActorOptions == null) {
+      rayActorOptions = new HashMap<>();
+    }
+    // TODO set runtime_env to rayActorOptions is not supported now.
+    ReplicaConfig replicaConfig = new ReplicaConfig(deploymentDef, initArgs, rayActorOptions);
+
+    deploymentConfig.setVersion(version);
+    deploymentConfig.setPrevVersion(prevVersion);
+
+    if (deploymentConfig.getAutoscalingConfig() != null
+        && deploymentConfig.getMaxConcurrentQueries()
+            < deploymentConfig.getAutoscalingConfig().getTargetNumOngoingRequestsPerReplica()) {
+      LOGGER.warn(
+          "Autoscaling will never happen, because 'max_concurrent_queries' is less than 'target_num_ongoing_requests_per_replica' now.");
+    }
+
+    boolean updating =
+        (boolean)
+            ((PyActorHandle) controller)
+                .task(
+                    PyActorMethod.of("deploy"),
+                    name,
+                    deploymentConfig.toProtoBytes(),
+                    replicaConfig.toProtoBytes(),
+                    routePrefix,
+                    Ray.getRuntimeContext().getCurrentJobId())
+                .remote()
+                .get();
+
+    String tag = "component=serve deployment=" + name;
+    if (updating) {
+      String msg = LogUtil.format("Updating deployment '{}'", name);
+      if (StringUtils.isNotBlank(version)) {
+        msg += LogUtil.format(" to version '{}'", version);
+      }
+      LOGGER.info("{}. {}", msg, tag);
+    } else {
+      LOGGER.info(
+          "Deployment '{}' is already at version '{}', not updating. {}", name, version, tag);
+    }
+
+    if (blocking) {
+      waitForDeploymentHealthy(name, -1);
+      String urlPart = url != null ? LogUtil.format(" at `{}`", url) : "";
+      LOGGER.info(
+          "Deployment '{}{}' is ready {}. {}",
+          name,
+          StringUtils.isNotBlank(version) ? "':'" + version : "",
+          urlPart,
+          tag);
+    }
+  }
+
+  /**
+   * Waits for the named deployment to enter "HEALTHY" status.
+   *
+   * <p>Raises RayServeException if the deployment enters the "UNHEALTHY" status instead or this
+   * doesn't happen before timeoutS.
+   *
+   * @param name
+   */
+  private void waitForDeploymentHealthy(String name, long timeoutS) {
```

Comment:
Can we change `long` to `Long`? Then, if `timeoutS` is `null` (instead of an arbitrary negative number), we can wait forever.

Reply:
Resolved.